Message broker payload updates

DFurnes opened the pull request:
#### Changes

Some lovely payload updates for the Message Broker. See #437. There are a few [open questions](https://github.com/DoSomething/voting-app/issues/437#issuecomment-156137621) that I posted in the issue, but putting this code up for preliminary review. :eyeglasses: 
#### Example Payloads

Brittany is a user from the US who gave us her email, but no cell phone. We'll send her some emails then:
<img width="617" alt="screen shot 2015-11-12 at 10 39 16 am" src="https://cloud.githubusercontent.com/assets/583202/11122661/297b269a-892a-11e5-96a8-0179513c8d6d.png">

Wilhelmina is a user from the US who gave us her email _and_ her cell phone. She therefore must love texting, so we'll send things to her phone:
<img width="348" alt="screen shot 2015-11-12 at 10 40 43 am" src="https://cloud.githubusercontent.com/assets/583202/11122684/3f5a9ba8-892a-11e5-89f9-9e45cefb28f9.png">

---

For review: @angaither @DeeZone

DFurnes committed:
Remove Mailchimp grouping ID and group name.

DFurnes committed:
Use newer Message Broker publish method.

DFurnes committed:
Add globalized email templates, and helpers for DS countries.

DFurnes committed:
Update Message Broker payload as specified in #437.

DFurnes committed:
Touchups to payload.

DFurnes committed:
Touchups to my messy code.

deezone commented:
@DFurnes In the case of example 2, if the app collects email and mobile please send both sets of info:
- `user_country`
- `user_language`

---
- `email`
  - `email_template`
  - `email_tags`
  - `email_list_id`
- `MERGE_VARS`

---
- `mobile`
  - `mobile_tags`
  - `opt_in_path_id`
- `MERGE_VARS`

In either or as a combo also send the `MERGE_VARS` values.

DFurnes committed:
Send all email or mobile fields if either is provided.

DFurnes committed:
Rename `opt_in_path_id` to `mobile_opt_in_path_id`.

DFurnes commented:
@DeeZone Done! Also addressed the feedback you provided in the issue.

deezone commented on .env.example:
```diff
@@ -31,9 +31,11 @@ REGISTER_TEMPLATE = mb-votingapp-signup
+MAILCHIMP_LIST_US = 555555
+MAILCHIMP_LIST_BR = 555555
+MAILCHIMP_LIST_MX = 555555
+MAILCHIMP_LIST_GLOBAL = 555555
```
+MAILCHIMP_LIST_US = f2fab1dfd4
+MAILCHIMP_LIST_BR = 66fd18c5a9
+MAILCHIMP_LIST_MX = 4f8e24d4f5
+MAILCHIMP_LIST_GLOBAL = 8e7844f6dd

DFurnes replied:
These are just example environment variables (since this app is able to be used for multiple campaigns). The actual environment variables are set in `.env` on the server. :)

deezone replied:
Can I leave it up to you to make the `.env` update?

DFurnes replied:
Yep, already did! Both Cats Gone Good and Celebs Gone Good (prod) have the appropriate environment variables for testing this once it gets merged and deployed.

deezone commented on app/Listeners/SendFirstVoteMessage.php:
```diff
@@ -30,37 +30,53 @@ public function handle(UserCastFirstVote $event)
             ];
+
+            // Send Mobile Commons opt-in path for US users, and MGage ID for international users.
+            $payload['mobile_opt_in_path_id'] = $event->user->country_code === 'US' ? env('MC_OPT_IN_PATH') : env('MGAGE_ID');
```
This needs to support `US`, `MX` and `BR`. `MX` and `BR` will have separate `MGAGE_ID`s. Perhaps:
- `MOBILE_OPT_IN_PATH_US`
- `MOBILE_OPT_IN_PATH_MX`
- `MOBILE_OPT_IN_PATH_BR`

As Message Broker will take care of using the appropriate service (Mobile Commons / mGage) but needs to know what opt_in path to add the user to.

DFurnes replied:
Does the application still need to send these, ~~or are they being stored on the Message Broker like MailChimp IDs~~? Update: perhaps a misunderstanding from a deleted comment.

DFurnes replied:
Okay, follow-up question – since we're sending `country_code` and the Message Broker is making the decision to use MobileCommons vs. MGage based on that, does it make sense for the application to just send all possible opt-in paths/list IDs identified by country code and let Message Broker decide between them?

deezone replied:
Ideally the app would take care of choosing what `MOBILE_OPT_IN_PATH` and `MAILCHIMP_LIST` to send the user to. Message Broker will take care of the connection to the account (API Key stuff).

DFurnes replied:
Got it. I updated it to send the appropriate `mobile_opt_in_path_id` and `mailchimp_list_id` for the user's country code in the payload.

deezone commented on app/Listeners/SendFirstVoteMessage.php:
```diff
@@ -30,37 +30,53 @@ public function handle(UserCastFirstVote $event)
             $payload['email'] = $event->user->email;
             $payload['subscribed'] = 1;
-            $payload['email_template'] = env('VOTE_TEMPLATE', 'mb-votingapp-vote');
+            $payload['email_template'] = env('VOTE_TEMPLATE', 'mb-votingapp-vote').'-'.normalize_country_code($event->user->country_code);
```
Does `normalize_country_code()` reduce the country code down to the supported countries? If so, please don't make the conversion as there's logic in the Message Broker app to deal with unsupported country code templates. I'd like the MB logic to be the _"one source of truth"_ on template selection.

DFurnes replied:
Oh, excellent! Will remove this.

DeeZone commented on app/Listeners/SendFirstVoteMessage.php:
```diff
@@ -30,37 +30,53 @@ public function handle(UserCastFirstVote $event)
-            ];
+
+            // Provide correct MailChimp list ID by country code
+            $mailchimpLists = config('services.message_broker.lists');
```
With the `user_country` in the payload all logic / references to `mailchimp_list_id` can be removed.

DFurnes replied:
So the list IDs are being stored in Message Broker?

DeeZone commented on app/Services/MessageBroker.php:
```diff
@@ -23,11 +23,10 @@ public function publishRaw($payload, $routingKey)
 
         $serializedPayload = serialize($payload);
-        $broker->publishMessage($serializedPayload);
+        $broker->publish($serializedPayload, $routingKey);
```
`$routingKey` needs to be set to `vote.event.CGG`.

DFurnes replied:
This is set [from an environment variable](https://github.com/DFurnes/voting-app/blob/c124986f71adb9794fdef8c737381309450e718b/.env.example#L25) so it can be environment-specific (e.g. `vote.event.CGG` here, but `vote.event.AGG` on the AGG site).

deezone commented on .env.example:
```diff
@@ -31,9 +31,11 @@ REGISTER_TEMPLATE = mb-votingapp-signup
 
-MAILCHIMP_GROUP_ID = 55555
-MAILCHIMP_GROUP_NAME = VotingApp
 MC_OPT_IN_PATH = 555555
```
See my comment below about:
- `MOBILE_OPT_IN_PATH_US`
- `MOBILE_OPT_IN_PATH_MX`
- `MOBILE_OPT_IN_PATH_BR`

@classicfreddie will provide the values.

DFurnes committed:
Don't normalize country codes.

Per feedback on #449, the email template will be normalized on the
Message Broker.

DFurnes committed:
Remove normalize function since we're no longer using.

DFurnes committed:
Add configuration variables for opt-in paths.

DFurnes committed:
Send correct MailChimp template and Opt-In Path by country.

DFurnes commented:
@DeeZone The latest round of changes is ready for re-review! :)

DFurnes commented:
Updated example payloads:

#### Email

Emailerina has provided her first name, birthdate, & email, and is from the US:
<img width="618" alt="screen shot 2015-11-13 at 11 23 29 am" src="https://cloud.githubusercontent.com/assets/583202/11151058/07904388-89f9-11e5-9114-39919acd7924.png">

#### Mobile

Mobilerina has provided her first name, birthdate, email, & mobile number, and is from the US:
<img width="617" alt="screen shot 2015-11-13 at 11 22 43 am" src="https://cloud.githubusercontent.com/assets/583202/11151050/03e859e6-89f9-11e5-8ef0-41ef46593598.png">

deezone commented:
:+1:

angaither commented:
:+1:
